post-til-virksomhet/src/main/java/no/difi/meldingsutveksling/ptv/CorrespondenceAgencyMessageFactory.java

```diff
@@ -0,0 +1,218 @@
+        // Name of the message recipient TODO: Skal hentes fra SR basert paa orgnr
+        tokens.getTextToken().add(createTextToken(2, msg.getEnvelope().getReceiver().getName()));
+
+        notification.setTextTokens(notificationFactory.createNotification2009TextTokens(tokens));
+
+        JAXBElement<ReceiverEndPointBEList> receiverEndpoints = createReceiverEndPoint();
+        notification.setReceiverEndPoints(receiverEndpoints);
+
+        // Should the user be allowed to forward the message from portal
+        correspondence.setAllowForwarding(objectFactory.createMyInsertCorrespondenceV2AllowForwarding(false));
+        // Name of the message sender, always "Avsender"
+        correspondence.setMessageSender(objectFactory.createMyInsertCorrespondenceV2MessageSender("Avsender"));
+
+        notifications.getNotification().add(notification);
+        correspondence.setNotifications(objectFactory.createMyInsertCorrespondenceV2Notifications(notifications));
+
+        no.altinn.services.serviceengine.correspondence._2009._10.ObjectFactory correspondenceObjectFactory = new no.altinn.services.serviceengine.correspondence._2009._10.ObjectFactory();
+        final InsertCorrespondenceV2 myInsertCorrespondenceV2 = correspondenceObjectFactory.createInsertCorrespondenceV2();
+        myInsertCorrespondenceV2.setCorrespondence(correspondence);
+        myInsertCorrespondenceV2.setSystemUserCode(systemUserCode);
+        // Reference set by the message sender
+        myInsertCorrespondenceV2.setExternalShipmentReference(msg.getEnvelope().getConversationId());
+
+        return myInsertCorrespondenceV2;
+    }
+
+    private static String queryPayload(PutMessageRequestWrapper message, String xpath) {
+        String result;
+
+        if(message.getMessageType() != PutMessageRequestWrapper.MessageType.EDUMESSAGE){
+            return "";
+        }
+
+        XPathFactory xPathFactory = XPathFactory.newInstance();
+        XPath xPath = xPathFactory.newXPath();
+        try {
+            XPathExpression expression = xPath.compile(xpath);
+            String doc;
+            if (message.getPayload() instanceof String) {
+                doc = (String) message.getPayload();
+                doc = unescapeHtml(doc);
+            } else {
+                doc = ((Node) message.getPayload()).getFirstChild().getTextContent().trim();
+            }
+
+            DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
+            DocumentBuilder builder = factory.newDocumentBuilder();
+            Document document = builder.parse(new ByteArrayInputStream(doc.getBytes(java.nio.charset.Charset.forName("utf-8"))));
+            result = expression.evaluate(document);
+        } catch (XPathExpressionException | ParserConfigurationException | SAXException | IOException e) {
+            throw new IllegalArgumentException("Could not execute query \'"+xpath+"\'  on the payload", e);
```

Comment:
Her må vi logge etter loggingregimet vårt. For å klare å feilsøke så må vi logge med endel andre ting for å kunne identifisere hvilket integrasjonspunkt som feiler i ELK stack, men også for å kunne feilsøke opp mot de parametrene meldingen sendes under.

Se på feks. IntegrasjonspunktImpl på bruk av Audit.error(...)